src/main/java/org/opensearch/neuralsearch/processor/combination/ScoreCombiner.java

```diff
@@ -59,32 +76,127 @@ private void combineShardScores(final ScoreCombinationTechnique scoreCombination
             scoreCombinationTechnique
         );
 
+        // If sorting is enabled then create the map of sort field values per docId
+        Map<Integer, Object[]> docIdSortFieldMap = null;
+        List<TopFieldDocs> topFieldDocs = null;
+        if (sort != null) {
+            final boolean isSortByScore = checkIfSortOrderByScore(sort);
+            topFieldDocs = new ArrayList<>();
+            for (int i = 0; i < topDocsPerSubQuery.size(); i++) {
+                if (topDocsPerSubQuery.get(i).scoreDocs.length != 0) {
+                    topFieldDocs.add((TopFieldDocs) topDocsPerSubQuery.get(i));
+                }
+            }
+            docIdSortFieldMap = getDocIdSortFieldsMap(compoundQueryTopDocs, isSortByScore, combinedNormalizedScoresByDocId);
+        }
+
         // - sort documents by scores and take first "max number" of docs
         // create a collection of doc ids that are sorted by their combined scores
-        List<Integer> sortedDocsIds = getSortedDocIds(combinedNormalizedScoresByDocId);
+        List<Integer> sortedDocsIds = getSortedDocIds(combinedNormalizedScoresByDocId, topFieldDocs, sort);
 
         // - update query search results with normalized scores
-        updateQueryTopDocsWithCombinedScores(compoundQueryTopDocs, topDocsPerSubQuery, combinedNormalizedScoresByDocId, sortedDocsIds);
+        updateQueryTopDocsWithCombinedScores(
+            compoundQueryTopDocs,
+            topDocsPerSubQuery,
+            combinedNormalizedScoresByDocId,
+            sortedDocsIds,
+            docIdSortFieldMap,
+            sort != null ? true : false
+        );
+    }
+
+    private boolean checkIfSortOrderByScore(Sort sort) {
+        if (sort != null) {
```

Comment:
better fail fast - start with case if sort == null then return, then the rest of code doesn't need to be in inside the if condition